File: ux/client/flogram/connector.js
// -*- web -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2021 all rights reserved


// externals
import React from 'react'
import { graphql, useFragment } from 'react-relay/hooks'
// local
import styles from './styles'


// render the connectors between factories and slots
export const Connector = (props) => {
    // pull the data
    const connector = useFragment(graphql`
        fragment connector_connector on Connector {
            id
            inp
            factoryAt {
                x
                y
            }
            slotAt {
                x
                y
            }
        }
    `, props.connector)

    // unpack
    // the direction flag
    const inp = connector.inp
    // the factory diagram coordinates
    const factoryAt = connector.factoryAt
    // and the slot coordinates
    const slotAt = connector.slotAt

    // handle inputs vs. outputs
    const delta = 2 * (inp ? -1 : 1)
    // compute the connector path
    const path = `
        M ${factoryAt.x + delta} ${factoryAt.y}
        L ${slotAt.x - delta} ${slotAt.y}
        L ${slotAt.x} ${slotAt.y}
        `

    // mix my paint
    const connectorStyle = { ...styles.connector, ...props.style }
    // render
    return (
        <path d={path} style={connectorStyle} />
    )
}


// end of file

File: pkg/gql/Connector.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2021 all rights reserved


# externals
import graphene

# field types
from .Position import Position


# a representation of a binding that links a slot to a factory
class Connector(graphene.ObjectType):
    """
    A connector from a product to a factory
    """

    # the fields
    id = graphene.ID()
    inp = graphene.Boolean(required=True)
    factoryAt = graphene.Field(Position, required=True)
    slotAt = graphene.Field(Position, required=True)


    # the resolvers
    def resolve_id(connector, *_):
        return connector.relay

    def resolve_inp(connector, *_):
        return connector.typename == "Input"

    def resolve_factoryAt(connector, *_):
        # unpack
        x, y = connector.factory.position
        # build a position and return it
        return Position(x=x, y=y)

    def resolve_slotAt(connector, *_):
        # unpack
        x, y = connector.slot.position
        # build a position and return it
        return Position(x=x, y=y)


# end of file
